etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/metadata/PartnerMetadataOrchestrator.java

```diff
@@ -15,45 +22,89 @@ public class PartnerMetadataOrchestrator {
+    public void updateMetadataForSentOrder(String receivedSubmissionId, String sentSubmissionId)
             throws PartnerMetadataException {
-        // call the metadata storage and add the sent order's submission ID to the existing metadata
-        // entry
-        // PartnerMetadata may need to be updated to store both the received order's submission ID
-        // _and_ the sent order's submission ID.
+
+        PartnerMetadata partnerMetadata =
+                partnerMetadataStorage.readMetadata(receivedSubmissionId).orElseThrow();
+        if (!sentSubmissionId.equals(partnerMetadata.sentSubmissionId())) {
+            partnerMetadata = partnerMetadata.withSentSubmissionId(sentSubmissionId);
+            partnerMetadataStorage.saveMetadata(partnerMetadata);
+        }
+
+        String receiver;
+        try {
+            String bearerToken = rsclient.getRsToken();
+            String responseBody = rsclient.requestHistoryEndpoint(sentSubmissionId, bearerToken);
+            receiver = getReceiverName(responseBody);
+        } catch (ReportStreamEndpointClientException | FormatterProcessingException e) {
+            throw new PartnerMetadataException(
+                    "Unable to retrieve metadata from RS history API", e);
+        }
+
+        partnerMetadata = partnerMetadata.withReceiver(receiver);
+        partnerMetadataStorage.saveMetadata(partnerMetadata);
     }
 
-    public Optional<PartnerMetadata> getMetadata(String submissionId)
+    public Optional<PartnerMetadata> getMetadata(String receivedSubmissionId)
             throws PartnerMetadataException {
         // call the metadata storage to get the metadata.
         // check if the receiver is filled out, and if it isn't, call the RS history API to get the
         // receiver.
         // if had to call the history API, extract the receiver and call the metadata storage to
         // save the metadata with the receiver added.
         // return the metadata.
-        return partnerMetadataStorage.readMetadata(submissionId);
+        return partnerMetadataStorage.readMetadata(receivedSubmissionId);
+    }
+
+    String getReceiverName(String responseBody) throws FormatterProcessingException {
+        Map<String, Object> responseObject =
+                formatter.convertJsonToObject(responseBody, new TypeReference<>() {});
+        Object destinationsObj = responseObject.get("destinations");
+        if (!(destinationsObj instanceof ArrayList<?> destinationsList)) {
+            throw new FormatterProcessingException(
+                    "destinations is not an ArrayList", new Exception());
+        }
+
+        if (destinationsList.isEmpty()
+                || !(destinationsList.get(0) instanceof Map<?, ?> destination)) {
+            throw new FormatterProcessingException(
+                    "First item in destinations is not a Map", new Exception());
+        }
+
+        var organizationId = destination.get("organization_id");
```

Comment:
Can we make both of these Strings? Just more explicit instead of relying on the jdk to figure out the type

Reply:
Will do